### ISSUE TITLE
Question about roll_timesteps in Trajectory.forward_test

### ISSUE THREAD

**Issue by xuzhuoran0106** [hustvl/DiffusionDrive#55]
Hi, This is an impressive work for introducing diffusion model  in E2E-AD. I would like to ask some questions.

ref core code
```
step_num = 2
step_ratio = 20 / step_num
roll_timesteps = (np.arange(0, step_num) * step_ratio).round()[::-1].copy().astype(np.int64)
roll_timesteps = torch.from_numpy(roll_timesteps).to(device)

for k in roll_timesteps[:]:
  ...
  timesteps = k # actually k is 10 and 0
  ...
  img = self.diffusion_scheduler.step(model_output=x_start, timestep=k, sample=img).prev_sample
  ...
```
In the code above, the roll_timesteps is actually [10, 0], for each self.diffusion_scheduler.step, second param timestep list is [10, 0].
Considering the.diffusion_scheduler.num_train_timesteps is 1000, the self.diffusion_scheduler.num_inference_steps is also 1000, the interval of timestep in inference stage is 1.
The code in DDIM.step for calculating prev_timestep is
'''

'''
The prev_timestep for the first denoise step (with k=10) is 9, so the denoised sample is at timestep 9. 
The second denoise step (with direct k=0), actually directs return the prediceted 'pose_reg' result.

My question, "The denoising step is executed 2 times, but the timestep is not consecutive. For example consecutive timestep is 10->9->8 in DDPM, or 10->8->6->4... in DDIM, with 'time_spaceing' interval."
Could you explain why the forward denoise is not a standard DDIM process? How much impact does it have on the results?

**Comment by TrayfourYu** [comment]
![Image](https://github.com/user-attachments/assets/5c015b53-a2b5-4cc5-9b9a-49114ed6497a)

**Comment by TrayfourYu** [comment]
In the https://huggingface.co/docs/diffusers/using-diffusers/write_own_pipeline

> scheduler.set_timesteps(50)
> scheduler.timesteps
> tensor([980, 960, 940, 920, 900, 880, 860, 840, 820, 800, 780, 760, 740, 720,
>     700, 680, 660, 640, 620, 600, 580, 560, 540, 520, 500, 480, 460, 440,
>     420, 400, 380, 360, 340, 320, 300, 280, 260, 240, 220, 200, 180, 160,
>     140, 120, 100,  80,  60,  40,  20,   0])

 The roll_timesteps is [10, 0], maybe just equal to `scheduler.set_timesteps(2)`.

**Comment by FlowEternal** [comment]
There are also some other issues that should be further clarified: 
(1) In this paper, the training T_trunc=50, which is the same as the code. But the T_truc in the forward test is T_trunc=8, which is not consistent with training. How much influence will this setting on the final pdms score is not clear.

(2) if the T_truc in the forward test is 8, then why the roll timestamp=(10,0). would it be more reasonable to set roll timestamp to be some thing like (8,4) and set inference step to be about 250(1000/250=4). then we start with timestamp 8, the model predict estimated final trajectory(actually offset) with timestamp 8, the DDIM step function then use the estimated final trajectory and initial trajectory to calculate direction and do interpolate with step 4(1000/250) to arrive at diffused trajectory at time stamp4(4=8-step=8-4) and then repeat the above process again to reach timestamp 0. 

(3) It seems that what matter most is the offset prediction to the anchor trajectory,Not the diffusion process. This trunced diffusion process seems unnecessary and the timestamp here does not matter at all.
 If the above insight hold, then it can explain why even setting roll timestamp=(10,0) it still work because every iter just do offset prediction and do not care timestamp at all.
That may also explain why this diffusion based decoder can not overfit on very small dataset because the added random diffusion noise prevent it from finding fixed offset pattern in small dataset.leading to traj loss Fluctuation

For example, I can design a decoder that purely predict offset to the original anchor trajectory and refine them one layer by another, which does not involve fancy diffusion and denoising process at all and can still expect good performance.In fact, it can be done by simple setting the noise in the code to zero and train as before. it is mathematically equivalent to do pure offset prediction, not involving any diffusion.

So what is the benefit of introducing diffusion since what matters is the offset prediction to anchor(like line detection using anchor based method).Maybe data augmentation? maybe no benefit at all. In my opinion the real innovation of this paper lie in the introduction of plan anchor with offset prediction not diffusion itself, whose real benefit is unknown

**Comment by ChuXX325** [comment]
> There are also some other issues that should be further clarified: (1) In this paper, the training T_trunc=50, which is the same as the code. But the T_truc in the forward test is T_trunc=8, which is not consistent with training. How much influence will this setting on the final pdms score is not clear.
> 
> (2) if the T_truc in the forward test is 8, then why the roll timestamp=(10,0). would it be more reasonable to set roll timestamp to be some thing like (8,4) and set inference step to be about 250(1000/250=4). then we start with timestamp 8, the model predict estimated final trajectory(actually offset) with timestamp 8, the DDIM step function then use the estimated final trajectory and initial trajectory to calculate direction and do interpolate with step 4(1000/250) to arrive at diffused trajectory at time stamp4(4=8-step=8-4) and then repeat the above process again to reach timestamp 0.
> 
> (3) It seems that what matter most is the offset prediction to the anchor trajectory,Not the diffusion process. This trunced diffusion process seems unnecessary and the timestamp here does not matter at all. If the above insight hold, then it can explain why even setting roll timestamp=(10,0) it still work because every iter just do offset prediction and do not care timestamp at all. That may also explain why this diffusion based decoder can not overfit on very small dataset because the added random diffusion noise prevent it from finding fixed offset pattern in small dataset.leading to traj loss Fluctuation
> 
> For example, I can design a decoder that purely predict offset to the original anchor trajectory and refine them one layer by another, which does not involve fancy diffusion and denoising process at all and can still expect good performance.In fact, it can be done by simple setting the noise in the code to zero and train as before. it is mathematically equivalent to do pure offset prediction, not involving any diffusion.
> 
> So what is the benefit of introducing diffusion since what matters is the offset prediction to anchor(like line detection using anchor based method).Maybe data augmentation? maybe no benefit at all. In my opinion the real innovation of this paper lie in the introduction of plan anchor with offset prediction not diffusion itself, whose real benefit is unknown
There doesn't seem to be a 1D Unet used in the code? If so, how is the diffusion model implemented?

<img width="2825" height="1021" alt="Image" src="https://github.com/user-attachments/assets/a8745f16-7a7a-4eb1-97f7-c6715fa65ccd" />

**Comment by ChuXX325** [comment]
> There are also some other issues that should be further clarified: (1) In this paper, the training T_trunc=50, which is the same as the code. But the T_truc in the forward test is T_trunc=8, which is not consistent with training. How much influence will this setting on the final pdms score is not clear.
> 
> (2) if the T_truc in the forward test is 8, then why the roll timestamp=(10,0). would it be more reasonable to set roll timestamp to be some thing like (8,4) and set inference step to be about 250(1000/250=4). then we start with timestamp 8, the model predict estimated final trajectory(actually offset) with timestamp 8, the DDIM step function then use the estimated final trajectory and initial trajectory to calculate direction and do interpolate with step 4(1000/250) to arrive at diffused trajectory at time stamp4(4=8-step=8-4) and then repeat the above process again to reach timestamp 0.
> 
> (3) It seems that what matter most is the offset prediction to the anchor trajectory,Not the diffusion process. This trunced diffusion process seems unnecessary and the timestamp here does not matter at all. If the above insight hold, then it can explain why even setting roll timestamp=(10,0) it still work because every iter just do offset prediction and do not care timestamp at all. That may also explain why this diffusion based decoder can not overfit on very small dataset because the added random diffusion noise prevent it from finding fixed offset pattern in small dataset.leading to traj loss Fluctuation
> 
> For example, I can design a decoder that purely predict offset to the original anchor trajectory and refine them one layer by another, which does not involve fancy diffusion and denoising process at all and can still expect good performance.In fact, it can be done by simple setting the noise in the code to zero and train as before. it is mathematically equivalent to do pure offset prediction, not involving any diffusion.
> 
> So what is the benefit of introducing diffusion since what matters is the offset prediction to anchor(like line detection using anchor based method).Maybe data augmentation? maybe no benefit at all. In my opinion the real innovation of this paper lie in the introduction of plan anchor with offset prediction not diffusion itself, whose real benefit is unknown

<img width="2825" height="1021" alt="Image" src="https://github.com/user-attachments/assets/aab3745d-ba5a-41fb-9346-cca1118df3a2" />
There doesn't seem to be a 1D Unet used in the code? If so, how is the diffusion model implemented?

**Comment by Chenfanqing** [comment]
> (3) It seems that what matter most is the offset prediction to the anchor trajectory,Not the diffusion process. This trunced diffusion process seems unnecessary and the timestamp here does not matter at all.


I share the same concerns. From the open-source code, it appears that:

1. In the forward train process, the pred trajectory is obtained by adding noise to the anchor, where the noise is interpreted as an offset. Thus, I  think the diffusion may not add much value, because if the goal is merely to predict offsets,  MLPs could suffice. The forward test seems to further support this point: it performs a complete forward train procedure to obtain the pred trajectory, and then applies diffusion step to the pred traj. What is the purpose of this step?

img = self.diffusion_scheduler.step(
model_output=x_start,
timestep=k,
sample=img
).prev_sample


2. From the diffusion perspective, the usual intuition in standard tasks is: add noise to the gt until it becomes pure noise, and then denoise the pure noise to generate the pred. By analogy in this framework, the code seems to add noise to the anchor to obtain the pred, and then denoise the pred, which would simply recover the anchor, wouldn’t it? What is the rationale or theoretical basis for doing this?
